### ISSUE TITLE
Rancher storage services

### ISSUE THREAD

**Issue by will-chan** [rancher/rancher#2574]
The ability for Rancher to launch storage services (Gluster, Ceph, Nexenta, etc.) via the catalog and support the following features:

- automatically create a Rancher storage pool from the storage service launched
- CRUD volumes from storage pool
- users can specify volumes from a specific storage pool either by the Rancher volume or via the combination of a Docker driver + volume name combination when launching a container.
- able to launch a container into hosts that belong to the same storage pool (so they access potentially access the same volume, for migration across hosts, or for HA of containers from a service)

**Comment by ebr** [comment]
+1 !!

**Comment by cjellick** [comment]
Container and volume functionality

Compose/API
- If a user specifies just a volume name in compose/api and a rancher-shared-volume of that name **DOES** exists, then that share volume should be used.
- If a user specifies just a volume name in compose/api and a rancher-shared-volume of that name **DOES NOT** exist, then a simple local volume should be created
- If a user specifies just a volume name in compose/api and more than one rancher-shared-volume of that name exists, then the container should fail to provision.
- If a user specifies a `volumeDriver=<user provided storage service name>` in compose/api, then the volumes specified in the request should be created as rancher-shared-volumes


Native Docker
- Container backpopulation should discover any rancher volumes associated with a container and create them in the right storage pool by matching the driver name of the volume to the driverName field on the storage pool
- When a user creates a volume via `docker volume create ...` rancher **will not** auto-discover that volume and create it in rancher. The same goes for deleting a volume.
  - In other words, the volume will only show up in Rancher when it is associated with a container and deleting a volume on the host will not be reflected in Rancher.

**Comment by cjellick** [comment]
UI changes ( Just my best guess ) 
- A user should be able to view their shared storage pools. Possible things they would want to see in this storage pool:
  - Hosts in the pool
  - Volumes in the pool
  - Containers in the pool
- A user should be able to list, create, and delete volumes in a storage pool
- Local storage pools (the one that is mapped 1:1 with a host) should not be displayed (I'm not sure that we have a way to distinguish these in the API other than the driverName being local...that could be good enough)

**Comment by cjellick** [comment]
i dont know how important the UI is (read that as I literally dont know, not that i think it is unimportant)
…you can have pretty complete functionality with the UI, but the UI would be the best way to show off what is going on

**Comment by cjellick** [comment]
Discussed UI briefly with @ibuildthecloud and @vincent99. For the short term, just want to a read-only UI for storage pools that shows the hosts and volumes in the pool.

Later need ability to delete a pool, but that needs tested.

**Comment by will-chan** [comment]
Closing issue as we have the feature in Experimental mode now.  All other issues/bugs/tasks/enhancements should be tracked separately.
